### PR TITLE
[#1895]Implement unit tests for pipeline element 'MergeByTimeProcessor'

### PR DESCRIPTION
**Pull request by rahulbiswas876**
#
  
  

### Purpose

see [1895](https://github.com/apache/streampipes/issues/1895)

### Remarks

PR introduces (a) breaking change(s): no

PR introduces (a) deprecation(s): no

### CODE REVIEW

File: streampipes-extensions/streampipes-processors-filters-jvm/src/main/java/org/apache/streampipes/processors/filters/jvm/processor/merge/MergeByTimeProcessor.java [hunk 1]
```diff
@@ -40,10 +40,10 @@
 
 public class MergeByTimeProcessor extends StreamPipesDataProcessor {
 
-  private static final String TIMESTAMP_MAPPING_STREAM_1_KEY = "timestamp_mapping_stream_1";
-  private static final String TIMESTAMP_MAPPING_STREAM_2_KEY = "timestamp_mapping_stream_2";
+  protected static final String TIMESTAMP_MAPPING_STREAM_1_KEY = "timestamp_mapping_stream_1";
+  protected static final String TIMESTAMP_MAPPING_STREAM_2_KEY = "timestamp_mapping_stream_2";
   private static final String NUMBER_MAPPING = "number_mapping";
```
Comment (bossenti) [inline streampipes-extensions/streampipes-processors-filters-jvm/src/main/java/org/apache/streampipes/processors/filters/jvm/processor/merge/MergeByTimeProcessor.java:45]: Would prefer to stay consistent here
So if we change from `private` to `protected`, then let's do it for all final String variables here

File: streampipes-extensions/streampipes-processors-filters-jvm/src/main/test/org/apache/streampipes/processors/filters/jvm/processor/merge/TestMergeByTimeProcessor.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            map.put(MergeByTimeProcessor.TIMESTAMP_MAPPING_STREAM_1_KEY, timestamp);
+        } else {
+            map.put(MergeByTimeProcessor.TIMESTAMP_MAPPING_STREAM_2_KEY, timestamp);
+        }
+        return EventFactory.fromMap(map,
+            new SourceInfo("test", streamId),
+            new SchemaInfo(null, Lists.newArrayList()));
+    }
+
+    class StoreEventCollector implements SpOutputCollector {
```
Comment (bossenti) [inline streampipes-extensions/streampipes-processors-filters-jvm/src/main/test/org/apache/streampipes/processors/filters/jvm/processor/merge/TestMergeByTimeProcessor.java:194]: See https://github.com/apache/streampipes/issues/1891#issuecomment-1760185990

Let's align here

**Comment by bossenti** [comment]
Thanks a lot for your contribution @rahulbiswas876 🙏🏼

File: streampipes-extensions/streampipes-processors-filters-jvm/src/main/java/org/apache/streampipes/processors/filters/jvm/processor/merge/MergeByTimeProcessor.java (hunk 1 above)
Comment (rahulbiswas876) [reply streampipes-extensions/streampipes-processors-filters-jvm/src/main/java/org/apache/streampipes/processors/filters/jvm/processor/merge/MergeByTimeProcessor.java]: addressed your comments

File: streampipes-extensions/streampipes-processors-filters-jvm/src/main/test/org/apache/streampipes/processors/filters/jvm/processor/merge/TestMergeByTimeProcessor.java (hunk 2 above)
Comment (rahulbiswas876) [reply streampipes-extensions/streampipes-processors-filters-jvm/src/main/test/org/apache/streampipes/processors/filters/jvm/processor/merge/TestMergeByTimeProcessor.java]: addressed your comment

**Comment by rahulbiswas876** [comment]
This is my first open source PR. I really enjoyed and learned while implementing the unit test how actually a good software code is written. Thanks team and @bossenti  reviewing the PR.

I am open to learn and excel. Please do let me know how I can improve my contributions.

File: .gitignore [hunk 3]
```diff
@@ -31,6 +31,7 @@
 .idea/dynamic.xml
 .idea/uiDesigner.xml
 .idea/*
+my-remarks.txt
```
Comment (bossenti) [inline .gitignore]: I guess this was committed by accident 🙂

Comment (rahulbiswas876) [reply .gitignore]: yeah I have fixed it alreay.

**Comment by bossenti** [comment]
There are some checkstyle issues (formatting issues): https://github.com/apache/streampipes/actions/runs/6504978848/job/17668735199?pr=2023#step:4:69572

**Comment by rahulbiswas876** [comment]
> There are some checkstyle issues (formatting issues): https://github.com/apache/streampipes/actions/runs/6504978848/job/17668735199?pr=2023#step:4:69572

can you please guide me which code style you guys follow.. Currently I am following GoogleStyle
image

**Comment by rahulbiswas876** [comment]
> There are some checkstyle issues (formatting issues): https://github.com/apache/streampipes/actions/runs/6504978848/job/17668735199?pr=2023#step:4:69572

fixed the check-style violations.

**Comment by bossenti** [comment]
> > There are some checkstyle issues (formatting issues): https://github.com/apache/streampipes/actions/runs/6504978848/job/17668735199?pr=2023#step:4:69572
> 
> can you please guide me which code style you guys follow.. Currently I am following GoogleStyle image

You can find our checkstyle config [here](https://github.com/apache/streampipes/tree/dev/tools/maven)

**Comment by bossenti** [comment]
Would probably be worth to mention somewhere in the documentation 🙈

**Comment by rahulbiswas876** [comment]
@bossenti can you please re-review the PR

**Comment by bossenti** [comment]
Hey @rahulbiswas876,

I really appreciate your motivation to contribute.
Please keep in mind that we are a community of volunteers trying to support and improve StreamPipes as much as possible. However, we all contribute to this project in our spare time. Therefore, it may take some time before we are able to respond to requests. This includes PR reviews. We won't forget about your PR, even if it might take us a few days to respond to updates.
I was pretty responsive this morning, but that can't be taken for granted.

File: streampipes-extensions/streampipes-processors-filters-jvm/src/main/test/org/apache/streampipes/processors/filters/jvm/processor/merge/TestMergeByTimeProcessor.java
Comment (bossenti) [inline streampipes-extensions/streampipes-processors-filters-jvm/src/main/test/org/apache/streampipes/processors/filters/jvm/processor/merge/TestMergeByTimeProcessor.java:1]: Can you please add a directory called `java` between `test` and `org`?
This allows the file to be recognized as a test case

**Comment by rahulbiswas876** [comment]
> Hey @rahulbiswas876,
> 
> I really appreciate your motivation to contribute. Please keep in mind that we are a community of volunteers trying to support and improve StreamPipes as much as possible. However, we all contribute to this project in our spare time. Therefore, it may take some time before we are able to respond to requests. This includes PR reviews. We won't forget about your PR, even if it might take us a few days to respond to updates. I was pretty responsive this morning, but that can't be taken for granted.

Hey @bossenti,

Since it is my 1st open source PR after struggling over the week to set up dev env and then learning git and all, I was overexcited and eager to get merged 🙈

But Yes, I acknowledge your feedback and thank the community for giving me the opportunity to learn and share openly.

**Review by bossenti — APPROVED** [review]

**Comment by bossenti** [comment]
> I was overexcited and eager to get merged

Keep that spirit @rahulbiswas876!
Just wanted to manage expectations 🙂

**Pull request merged**